src/Tmds.Ssh/SshAgent.cs

```diff
@@ -23,60 +23,77 @@ sealed class SshAgent : IDisposable
-
         public struct Identity
         {
             public string Comment { init; get; }
             public byte[] PublicKey { init; get; }
         }
 
-        public static EndPoint? DefaultEndPoint
+        public static string? DefaultEndPoint
         {
             get
             {
-                string? sshAuthSock = Environment.GetEnvironmentVariable("SSH_AUTH_SOCK");
-
-                if (OperatingSystem.IsWindows())
-                {
-                    return null;
-                }
-
-                if (string.IsNullOrEmpty(sshAuthSock))
-                {
-                    return null;
-                }
-
-                _defaultEndPoint ??= new UnixDomainSocketEndPoint(sshAuthSock);
-
-                return _defaultEndPoint;
+                return OperatingSystem.IsWindows()
+                    ? "openssh-ssh-agent"
+                    : Environment.GetEnvironmentVariable("SSH_AUTH_SOCK");
             }
         }
 
-        private readonly EndPoint _endPoint;
+        private readonly string _endPoint;
         private readonly SequencePool _sequencePool;
 
-        private SocketSshConnection? _agentConnection;
+        private StreamSshConnection? _agentConnection;
 
-        public SshAgent(EndPoint endPoint, SequencePool sequencePool)
+        public SshAgent(string endPoint, SequencePool sequencePool)
         {
             _endPoint = endPoint;
             _sequencePool = sequencePool;
         }
 
         public async ValueTask<bool> TryConnect(CancellationToken cancellationToken)
         {
+            NamedPipeClientStream? pipe = null;
             Socket? socket = null;
             try
             {
-                socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
-                await socket.ConnectAsync(_endPoint, cancellationToken).ConfigureAwait(false);
-                _agentConnection = new SocketSshConnection(NullLoggerFactory.Instance.CreateLogger<SshClient>(), _sequencePool, socket);
+                var logger = NullLoggerFactory.Instance.CreateLogger<SshClient>();
+
+                if (OperatingSystem.IsWindows())
+                {
+                    // There is no easy way to detect if a named pipe exists or
+                    // not. File.Exists is the fastest but will actually open a
+                    // connection and close it. Using this won't open the pipe
+                    // Directory.GetFiles("\\\\.\\pipe\\", endPoint).Length > 0
+                    // But it is slower.
+                    if (!File.Exists(@$"\\.\pipe\{_endPoint}"))
+                    {
+                        return false;
+                    }
+
+                    pipe = new NamedPipeClientStream(
+                        ".",
+                        _endPoint,
+                        PipeDirection.InOut,
+                        PipeOptions.Asynchronous,
+                        System.Security.Principal.TokenImpersonationLevel.Anonymous);
```

Comment:
I'm actually surprised this works, I need to dig into the ssh agent implementation for Windows to actually see what they are doing but this would indicate to me that the keys aren't protected per user but are machine wide which is a bit worrisome.

Reply:
I looked at the source code. Surprisingly it gets the PID of the client who has connected to the pipe and uses that to open the process token and get the user information that way.

https://github.com/PowerShell/openssh-portable/blob/0c3137f62115857cabb6520c2a06ec59e854a56e/contrib/win32/win32compat/ssh-agent/agent.c#L304-L310

This means we can connect with `Anonymous` as the server won't try to impersonate the client pipe thread but rather go through the PID.